peerstore_test.go

```diff
@@ -248,24 +280,76 @@ func TestPeerstoreProtoStore(t *testing.T) {
 	}
 }
 
-func TestBasicPeerstore(t *testing.T) {
-	ps := NewPeerstore()
+func TestPeerstoreProtoStore(t *testing.T) {
+	runTestWithPeerstores(t, testAddrStreamDuplicates)
+}
 
+func testBasicPeerstore(t *testing.T, ps Peerstore) {
 	var pids []peer.ID
 	addrs := getAddrs(t, 10)
-	for i, a := range addrs {
-		p := peer.ID(fmt.Sprint(i))
+	for _, a := range addrs {
+		priv, _, _ := crypto.GenerateKeyPair(crypto.RSA, 512)
+		p, _ := peer.IDFromPrivateKey(priv)
 		pids = append(pids, p)
 		ps.AddAddr(p, a, PermanentAddrTTL)
 	}
 
 	peers := ps.Peers()
 	if len(peers) != 10 {
-		t.Fatal("expected ten peers")
+		t.Fatal("expected ten peers, got", len(peers))
 	}
 
 	pinfo := ps.PeerInfo(pids[0])
 	if !pinfo.Addrs[0].Equal(addrs[0]) {
 		t.Fatal("stored wrong address")
 	}
 }
+
+func TestBasicPeerstore(t *testing.T) {
+	runTestWithPeerstores(t, testBasicPeerstore)
+}
+
+func BenchmarkBasicPeerstore(b *testing.B) {
+	ps := NewPeerstore()
+
+	addrs := make(chan *peerpair, 100)
+
+	go addressProducer(b, addrs, 50000)
+
+	for {
+		pp, ok := <-addrs
+		if !ok {
+			break
+		}
+		pid := peer.ID(pp.ID)
+		ps.AddAddr(pid, pp.Addr, PermanentAddrTTL)
+	}
+}
+
+func benchmarkPeerstoreRateLimited(ps Peerstore) func(*testing.B) {
+	return func(b *testing.B) {
+		producer := make(chan *peerpair, 100)
+		addrs := make(chan *peerpair, 100)
+
+		go rateLimitedAddressProducer(b, addrs, producer, 60000, 100*time.Microsecond, 100*time.Microsecond)
+
+		b.ResetTimer()
+		for {
+			pp, ok := <-addrs
+			if !ok {
+				break
+			}
+			pid := peer.ID(pp.ID)
+			ps.AddAddr(pid, pp.Addr, PermanentAddrTTL)
+		}
+	}
+}
+
+func BenchmarkBasicPeerstoreRateLimited(b *testing.B) {
+	ps := NewPeerstore()
+	b.Run("PeerstoreBasic", benchmarkPeerstoreRateLimited(ps))
+
+	dsps, closer := setupDatastorePeerstore(b)
+	defer closer()
+	b.Run("PeerstoreDatastore", benchmarkPeerstoreRateLimited(dsps))
```

Comment:
Running benchmarks for less than 13s (though 15s seems the best) still results in nonsense results on this test:
```
$ go test -run none -bench RateLimited -benchtime 12s -benchmem
     | goos: darwin
     | goarch: amd64
     | pkg: github.com/libp2p/go-libp2p-peerstore
     | BenchmarkBasicPeerstoreRateLimited/PeerstoreBasic-8         	1000000000	         8.39 ns/op	       0 B/op	       0 allocs/op
     | BenchmarkBasicPeerstoreRateLimited/PeerstoreDatastore-8      10	         2302502401 ns/op	       33663786 B/op	       732672 allocs/op
```
```
$ go test -run none -bench RateLimited -benchtime 15s -benchmem
     | goos: darwin
     | goarch: amd64
     | pkg: github.com/libp2p/go-libp2p-peerstore
     | BenchmarkBasicPeerstoreRateLimited/PeerstoreBasic-8         	2000000000	         4.12 ns/op	       0 B/op	       0 allocs/op
     | BenchmarkBasicPeerstoreRateLimited/PeerstoreDatastore-8     	1000000000	        16.2 ns/op	       0 B/op	       0 allocs/op
```

Maybe a comment on the benchmark or a `make` rule that runs the benchmarks for an appropriate amount of time?

Reply:
🤦‍♂️ I just realised what is going wrong, the for loops in the benchmarks don't use `b.N` so the pattern for benchmarks is the following:

```go
func Benchmark...(b *testing.B) {
    // Do setup stuff
    
    // If setup is expensive, reset timer
    for i := 0; i < b.N; i++ { 
        // run function you want to benchmark
    }
}
```

`b.N` is how the testing pkg determines whether it has run the benchmark for long enough

Sorry @bigs, I should have caught that earlier. So I tested this on the previous commit and it works well

Reply:
The `BenchmarkBasicPeerstore` is also missing the `b.N` for loop.

Reply:
So I figured out a way to isolate the `peerstore.AddAddr` function in the benchmarks utilizing `b.StartTimer()` and `b.StopTimer()`, sorry for the large snippet: 

```go
func benchmarkBasicPeerstore(ps Peerstore) func(*testing.B) {
	return func(b *testing.B) {
		b.StopTimer()
		addrs := make(chan *peerpair, 100)

		go addressProducer(b, addrs, 50000)

		for i := 0; i < b.N; i++ {
			pp, ok := <-addrs
			if !ok {
				break
			}
			pid := peer.ID(pp.ID)
			b.StartTimer()
			ps.AddAddr(pid, pp.Addr, PermanentAddrTTL)
			b.StopTimer()
		}
	}
}

func BenchmarkBasicPeerstore(b *testing.B) {
	ps := NewPeerstore()
	b.Run("PeerstoreBasic", benchmarkBasicPeerstore(ps))

	dsps, closer := setupDatastorePeerstore(b)
	defer closer()
	b.Run("PeerstoreDatastore", benchmarkBasicPeerstore(dsps))
}

func benchmarkPeerstoreRateLimited(ps Peerstore) func(*testing.B) {
	return func(b *testing.B) {
		b.StopTimer()
		producer := make(chan *peerpair, 100)
		addrs := make(chan *peerpair, 100)

		go rateLimitedAddressProducer(b, addrs, producer, 60000, 100*time.Microsecond, 100*time.Microsecond)

		for i := 0; i < b.N; i++ {
			pp, ok := <-addrs
			if !ok {
				break
			}
			pid := peer.ID(pp.ID)
			b.StartTimer()
			ps.AddAddr(pid, pp.Addr, PermanentAddrTTL)
			b.StopTimer()
		}
	}
}

// BenchmarkBasicPeerstoreRateLimited is the same
```

The main win here is that `BenchmarkBasicPeerstoreRateLimited/PeerstoreBasic-8` doesn't appear slow waiting for address off the `addrs` chan. And it represents the memory profile more accurately as well.

Reply:
omg thx! the benchmarking docs are a bit spotty. this helps a lot. feel free to push a patch, otherwise, i’m happy to do it tomorrow!